### PR TITLE
Fix Json verification

### PR DESCRIPTION
**Pull request by stefan-aws**
By submitting this pull request, I confirm that my contribution is made under the terms of the MIT license.

### CODE REVIEW

File: src/JSON/Utils/Cursors.dfy [hunk 1]
```diff
@@ -142,6 +142,15 @@ module {:options "-functionSyntax:4"} JSON.Utils.Cursors {
     function Split() : (sp: Split<View>) requires Valid?
       ensures sp.SplitFrom?(this, (v: View) => v.Bytes())
       ensures beg != point ==> sp.StrictlySplitFrom?(this, (v: View) => v.Bytes())
+      ensures !BOF? ==> (sp.StrictlySplitFrom?(this, (v: View) => v.Bytes()) && sp.cs.StrictSuffixOf?(this))
```
Comment (atomb) [inline src/JSON/Utils/Cursors.dfy:145]: Given how much you're adding to the postconditions, would it be feasible to make `Split` opaque, as well?

File: src/JSON/Utils/Cursors.dfy [hunk 2]
```diff
@@ -265,6 +274,11 @@ module {:options "-functionSyntax:4"} JSON.Utils.Cursors {
       decreases SuffixLength()
       ensures ps.AdvancedFrom?(this)
       ensures !EOF? ==> ps.StrictlyAdvancedFrom?(this)
+      ensures ps.s == this.s
```
Comment (atomb) [inline src/JSON/Utils/Cursors.dfy:277]: Similar comment as for `Split`.

File: src/JSON/ZeroCopy/Deserializer.dfy [hunk 3]
```diff
@@ -70,16 +79,24 @@ module {:options "-functionSyntax:4"} JSON.ZeroCopy.Deserializer {
...
       var SP(before, cs) := WS(cs);
       var SP(val, cs) := cs.SkipByte().Split();
       var SP(after, cs) := WS(cs);
       SP(Grammar.Structural(before, val, after), cs)
     }
 
-    type ValueParser = sp: SubParser<Value> |
-        forall t :: sp.spec(t) == Spec.Value(t)
-      witness *
+    ghost predicate ValueParserValid(sp: SubParser<Value>) {
```
Comment (atomb) [inline src/JSON/ZeroCopy/Deserializer.dfy:95]: This seems like a good candidate to make opaque, as well, given the quantifier, though opacity in subtypes can be tricky.

File: src/JSON/Utils/Cursors.dfy (hunk 2 above)
Comment (stefan-aws) [reply src/JSON/Utils/Cursors.dfy]: Removed most ensures clauses.

File: src/JSON/Utils/Cursors.dfy (hunk 1 above)
Comment (stefan-aws) [reply src/JSON/Utils/Cursors.dfy:145]: Removed most ensures clauses.

File: src/JSON/ZeroCopy/Deserializer.dfy (hunk 3 above)
Comment (stefan-aws) [reply src/JSON/ZeroCopy/Deserializer.dfy:83]: Unfortunately turning the predicate opaque caused a couple of issues. For now, I propose we keep it as it is.

File: src/JSON/ConcreteSyntax.SpecProperties.dfy [hunk 4]
```diff
@@ -12,23 +12,20 @@ module {:options "-functionSyntax:4"} JSON.ConcreteSyntax.SpecProperties
...
-      | && (forall d | d < bracketed :: pd0.requires(d))
-      && (forall d | d < bracketed :: pd1.requires(d))
-      && (forall d | d < bracketed :: pd0(d) == pd1(d))
-    {
-      calc {
-        Spec.Bracketed(bracketed, pd0);
-        { ConcatBytes_Morphism(bracketed.data, pd0, pd1); }
-        Spec.Bracketed(bracketed, pd1);
-      }
+    calc {
```
Comment (atomb) [inline src/JSON/ConcreteSyntax.SpecProperties.dfy:25]: Nice simplification!

**Review by atomb — DISMISSED** [review]
Looks nice to me!

File: src/JSON/ZeroCopy/Deserializer.dfy [hunk 5]
```diff
@@ -259,9 +278,19 @@ module {:options "-functionSyntax:4"} JSON.ZeroCopy.Deserializer {
...
+        && ((!cs.BOF? || !cs.EOF?) && (s0 == SEPARATOR as opt_byte) ==> (var sp: Split<Structural<jcomma>> := sp; sp.cs.StrictSuffixOf?(cs)))
+        && ((s0 == SEPARATOR as opt_byte) ==> var sp: Split<Structural<jcomma>> := sp; sp.SplitFrom?(cs, st => Spec.Structural(st, SpecView)))
+        && ((!cs.BOF? || !cs.EOF?) && (s0 == CLOSE as opt_byte) ==> (var sp: Split<Structural<jclose>> := sp; sp.cs.StrictSuffixOf?(cs)))
+        && ((s0 == CLOSE as opt_byte) ==> var sp: Split<Structural<jclose>> := sp; sp.SplitFrom?(cs, st => Spec.Structural(st, SpecView)))
+    {}
+
     // The implementation and proof of this function is more painful than
     // expected due to the tail recursion.
-    function {:opaque} {:tailrecursion} Elements(
+    function {:timeLimit 30} {:vcs_split_on_every_assert} {:opaque} {:tailrecursion} Elements(
```
Comment (robin-aws) [inline src/JSON/ZeroCopy/Deserializer.dfy:293]:
```suggestion
    function {:vcs_split_on_every_assert} {:opaque} {:tailrecursion} Elements(
```

(And likewise below)

I assume this was here for debugging earlier. We should make sure we only include very generous wall-clock timeouts to maximize build reproducibility, and I don't see a huge value in bounding this more tightly (we could set an `:rlimit` instead if that's the intention)

File: src/JSON/ZeroCopy/Deserializer.dfy [hunk 6]
```diff
@@ -276,49 +305,110 @@ module {:options "-functionSyntax:4"} JSON.ZeroCopy.Deserializer {
...
-          assert elem.StrictlySplitFrom?(elems.cs, ElementSpec);
-        }
-        var elems' := AppendLast(open.cs, json, elems, elem, sep);
-        assert elems'.SplitFrom?(open.cs, SuffixedElementsSpec) by {
-          assert elems'.StrictlySplitFrom?(open.cs, SuffixedElementsSpec);
-        }
-        var bracketed := BracketedFromParts(cs0, open, elems', sep);
-        assert bracketed.StrictlySplitFrom?(cs0, BracketedSpec);
-        Success(bracketed)
+      if elem.cs.EOF? then
```
Comment (robin-aws) [inline src/JSON/ZeroCopy/Deserializer.dfy:308]: Just for my understanding, was the key improvement re-ordering the cases or was this a deeper optimization of some kind?

File: src/JSON/ZeroCopy/Deserializer.dfy [hunk 7]
```diff
@@ -426,19 +517,82 @@ module {:options "-functionSyntax:4"} JSON.ZeroCopy.Deserializer {
           assert SP(arr, cs').StrictlySplitFrom?(cs, Spec.Array);
           Spec.UnfoldValueArray(v);
         }
+        assert sp.StrictlySplitFrom?(cs, Spec.Value);
         Success(sp)
       else if c == '\"' as opt_byte then
-        var SP(str, cs) :- Strings.String(cs);
-        Success(SP(Grammar.String(str), cs))
+        var SP(str, cs') :- Strings.String(cs);
```
Comment (robin-aws) [inline src/JSON/ZeroCopy/Deserializer.dfy:523]: Worth a comment about why all the expansion/duplication is necessary here?

**Review by robin-aws — CHANGES_REQUESTED** [review]
Very nice improvements! Just the one blocking issue with `:timeLimit` from me

File: src/JSON/ZeroCopy/Deserializer.dfy (hunk 6 above)
Comment (stefan-aws) [reply src/JSON/ZeroCopy/Deserializer.dfy:308]: The case split is new I believe. As you can see from the Lemma `AboutTryStructural` I added, it is important to know that `!elemen.cs.EOF?`.

File: src/JSON/ZeroCopy/Deserializer.dfy (hunk 7 above)
Comment (stefan-aws) [reply src/JSON/ZeroCopy/Deserializer.dfy:523]: The cases have been there previously. I added the calc's such that post-condition of `Value` goes through. Each of the proofs requires prior knowledge about the shape of `c`, to use either `Grammar.String` or `Grammar.Bool`, and either `var f := _ => TRUE;` or `var f := _ => FALSE;` etc. I think there is not much duplication we can remove here, unfortunately.

File: src/JSON/ZeroCopy/Deserializer.dfy (hunk 6 above)
Comment (stefan-aws) [reply src/JSON/ZeroCopy/Deserializer.dfy:308]: I moved the Lemma `AboutTryStructural` to the actual `assertion by` clauses it is needed for.

File: src/JSON/ZeroCopy/Deserializer.dfy (hunk 5 above)
Comment (stefan-aws) [reply src/JSON/ZeroCopy/Deserializer.dfy:293]: You are right, it's been used for debugging. I'll remove it everywhere.

**Review by atomb — APPROVED** [review]
Looks good!

**Review by robin-aws — APPROVED** [review]
🚀

**Pull request merged**